exonum/src/runtime/mailbox/mod.rs

```diff
@@ -0,0 +1,126 @@
+// Copyright 2019 The Exonum Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! The module containing interfaces to request changes in the blockchain structure.
+
+mod blockchain_secretary;
+
+pub use blockchain_secretary::{BlockchainSecretary, MailboxContext};
+
+use std::cell::RefCell;
+
+use crate::runtime::{ArtifactId, ConfigChange, InstanceId, InstanceSpec};
+
+/// Optional callback to be called after request is completed by the blockchain core.
+pub type AfterRequestCompleted = Option<Box<dyn FnOnce() + 'static>>;
```

Comment:
Synchronous approach won't work good for heavy-weight artifacts in the future, since it will block core main thread.

Regarding node restarts: I think it's okay for now, since if node will be restarted during the deployment, it won't be deployed anyway (since it won't be registered in the dispatcher schema) and overall deploy will also fail. The whole deploy system currently is not designed to be bulletproof. That's a problem, yeah, but to resolve it we should do a research first (which will create a near-complete list of existing flaws and propose a generic algorithm to deal with them).